server/src/main/java/org/elasticsearch/action/SingleResultDeduplicator.java

```diff
@@ -77,17 +79,27 @@ private void doExecute(ActionListener<T> listener) {
                     waitingListeners = new ArrayList<>();
                 }
             }
-            doExecute(new ActionListener<>() {
-                @Override
-                public void onResponse(T response) {
-                    ActionListener.onResponse(listeners, response);
-                }
 
-                @Override
-                public void onFailure(Exception e) {
-                    ActionListener.onFailure(listeners, e);
-                }
-            });
+            // Create a child threadContext so that the parent context remains unchanged when the child execution returns.
+            // This ensures the parent listener does not see response headers from the child execution.
+            try (var ignore = threadContext.newRestorableContext(false).get()) {
+                // Restore the parent's original threadContext before proceed with the child execution.
+                // This ensures all executions, parent or child, always begin execution with the identical threadContext.
+                // This identical threadContext is the context of the first listener gets executed.
+                originalContext.restore();
+                doExecute(originalContext, new ActionListener<>() {
```

Comment:
I think this still leads to some surprises. If listener L1 triggers the first execution and then L2 comes along while L1 is still running, then when L1 completes we'll run the action again for L2 but that second run is happening _in the thread context of L1_. That seems trappy, I would expect the action to be executed in the context of one of the waiting listeners, not some long-obsolete context for an older listener.

Reply:
I thought always using the first listener's context is more predicatable. But I can see it being argued both ways. So I'll changed it as you suggested.